test/e2e/scheduling/framework.go

```diff
@@ -51,22 +54,57 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
 			framework.Failf("Timed out after %v waiting for stable cluster.", timeout)
 			break
 		}
-		allScheduledPods, allNotScheduledPods = getFilteredPods(c, masterNodes, metav1.NamespaceAll)
+		allScheduledPods, allNotScheduledPods = getScheduledAndUnscheduledPods(c, masterNodes, metav1.NamespaceAll)
 	}
 	return len(allScheduledPods)
 }
 
-// getFilteredPods lists scheduled and not scheduled pods in the given namespace, with succeeded and failed pods filtered out.
-func getFilteredPods(c clientset.Interface, masterNodes sets.String, ns string) (scheduledPods, notScheduledPods []v1.Pod) {
+// WaitForPodsToBeDeleted waits until pods that are terminating to get deleted.
+func WaitForPodsToBeDeleted(c clientset.Interface) {
+	startTime := time.Now()
+	deleting := getDeletingPods(c, metav1.NamespaceAll)
+	for len(deleting) != 0 {
+		if startTime.Add(timeout).Before(time.Now()) {
+			framework.Logf("Pods still not deleted")
+			for _, p := range deleting {
+				framework.Logf("%v/%v", p.Namespace, p.Name)
+			}
+			framework.Failf("Timed out after %v waiting for pods to be deleted", timeout)
+			break
+		}
+		time.Sleep(wait)
+		deleting = getDeletingPods(c, metav1.NamespaceAll)
+	}
+}
+
+// getScheduledAndUnscheduledPods lists scheduled and not scheduled pods in the given namespace, with succeeded and failed pods filtered out.
+func getScheduledAndUnscheduledPods(c clientset.Interface, masterNodes sets.String, ns string) (scheduledPods, notScheduledPods []v1.Pod) {
 	pods, err := c.CoreV1().Pods(ns).List(metav1.ListOptions{})
 	framework.ExpectNoError(err, "listing all pods in kube-system namespace while waiting for stable cluster")
 	// API server returns also Pods that succeeded. We need to filter them out.
 	filteredPods := make([]v1.Pod, 0, len(pods.Items))
-	for _, pod := range pods.Items {
-		if pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed {
-			filteredPods = append(filteredPods, pod)
+	for _, p := range pods.Items {
+		if !podTerminated(p) {
+			filteredPods = append(filteredPods, p)
 		}
 	}
 	pods.Items = filteredPods
 	return e2epod.GetPodsScheduled(masterNodes, pods)
 }
+
+// getDeletingPods returns whether there are any pods marked for deletion.
+func getDeletingPods(c clientset.Interface, ns string) []v1.Pod {
+	pods, err := c.CoreV1().Pods(ns).List(metav1.ListOptions{})
+	framework.ExpectNoError(err, "listing all pods in kube-system namespace while waiting for pods to terminate")
+	var deleting []v1.Pod
+	for _, p := range pods.Items {
+		if p.ObjectMeta.DeletionTimestamp != nil && !podTerminated(p) {
```

Comment:
just to double check my understanding, we could remove the ```&& !podTerminted``` condition and we would still be ok, right?

Reply:
could a pod have a deletiontimestamp set and not yet be in the succeeded/failed state? In that case it would still affect the pod counts

Reply:
I'm not entirely sure. I imagine different controllers (kubelet or apiserver) might update the fields or remove the pods at different times.